### PR TITLE
[FSDP] Implement local_state_dict and load_local_state_dict

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #72469

1. Implement the framework to allow user to choose among `state_dict`, `local_state_dict`, and `sharded_state_dict`.
2. Implement ShardedTensor compatible local_state_dict() and load_local_state_dict().

Differential Revision: [D33919683](https://our.internmc.facebook.com/intern/diff/D33919683/)

### CODE REVIEW

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:712]: Should we use some sort of basic dispatch mechanism to deconvolute / separate out implementation details from this code? 

Maybe something like 3 functions:

`def _local_post_state_dict_hook`
`def _full_post_state_dict_hook`
`def _sharded_state_dict_hook`

And in `_post_state_dict_hook`, we have:
```
dispatch: Dict[StateDictType, Callable] = { 
    StateDictType.LOCAL: _local_post_state_dict_hook,
    # and so on
}
```

and the implementation is just `return dispatch[state_dict_type](args)`

eventually, those 3 functions might be able to be moved into some sort of utils folder.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:728]: do we need to have `flat_param is not None` assert before this line instead of on 730?

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:739]: nice!

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:727]: nit: add some docs/description describing what the hook does?

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:761]: Why do we need to manually call this if it is registered as a hook in FSDP ctor?

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:775]: will `super().state_dict()` work and automatically call all the hooks?

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:790]: same comment around the dispatch idea as for `_post_state_dict_hook`.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:804]: Can we separate out any of this logic into helper functions? Such as the process to reconstruct the sharded parameter from `shards` object. 

Also, same comment around adding comments/documentation around what this code is doing for better maintainability / understanding.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:833]: we use `super()` call here but don't use it in `state_dict` implementation where we actually return `FullyShardedDataParallel.state_dict()`, that seems inconsistent and can cause issues/confusion?

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:685]: seems only used in FSDP code and users will call things like `local_state_dict`, should we prefix the impl with `_`?

File: test/distributed/fsdp/test_fsdp_state_dict.py
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_state_dict.py:40]: Wonder if we should test a more complicated set of models. A few ideas:

1. Nested FSDP module
2. FSDP module where shards need padding, i.e. total # of params is not divisible by the world size and we pad the shards. Seems like there's some specific logic there that we'd like to verify.
3. More complex module such as those with shared params, Transformer module, etc. These can be added as part of FSDP vs DDP parity test as I mentioned in another comment.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:792]: Logic for both local_state_dict loading and saving hooks seem nontrivial and I think they're probably worth testing as standalone functionality, which would save a bunch of time debugging if something in local_state_dict breaks down the line. What do you think?

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:1279]: Seems unrelated, why do we need to add `_lazy_init` in _rebuild_full_params now?

**Review by rohan-varma — COMMENTED** [review]
This is amazing work! Please see comments inline.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (zhaojuanmao) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:756]: if torch.cuda.is_available():
            torch.cuda.synchronize()

File: test/distributed/fsdp/test_fsdp_state_dict.py
Comment (fegin) [reply test/distributed/fsdp/test_fsdp_state_dict.py:113]: I remove the test for full_state_dict and sharded_state_dict. We should add the test when landing either full_state_dict or sharded_state_dict.

File: test/distributed/fsdp/test_fsdp_state_dict.py
Comment (fegin) [reply test/distributed/fsdp/test_fsdp_state_dict.py:127]: Add a TODO.

File: torch/distributed/fsdp/flatten_params_wrapper.py
Comment (fegin) [reply torch/distributed/fsdp/flatten_params_wrapper.py:43]: @zhaojuanmao FlattenParamsWrapper._fpw_module can be another FSDP. If we do not apply the _replace_by_prefix, we may get keys like: {"flat_param", "_fpw_module.inner.flat_param"}. The former key is the  flat_param of the outer/parent module and the later key is the flat_param of the inner/child module.

File: torch/distributed/fsdp/flatten_params_wrapper.py
Comment (fegin) [reply torch/distributed/fsdp/flatten_params_wrapper.py:434]: This is flat_state_dict() that is called when parameters are flattened. We do not override `state_dict()`, so self.state_dict(*args, **kwargs) is equivalent to  super().state_dict().

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (fegin) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:690]: Yes. Now that I added the check of is_root property, I remove the comment but still keep the check inside the API.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (fegin) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:775]: `self.state_dict()` makes more sense. I changed it.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (fegin) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:833]: local_state_dict() and sharded_state_dict() should all call `self.state_dict`. And self.state_dict has to call super().state_dict(). 

local_state_dict() and sharded_state_dict() are just wrapper to set state_dict_type first then call self.state_dict. The full state_dict does not have to do so because the state_dict_type defaults to STATE_DICT.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (fegin) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:685]: This should be exported. Uses may not call `local_state_dict` directly. In fact, users cannot call `local_state_dict` directly if the FSDP module is wrapped inside another non-FSDP module, e.g., Pyper models. This allow users to do something like:
with state_dict_type(LOCAL_STATE_DICT):
       pyper_model.state_dict()

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (fegin) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:739]: Correct, as long as `from_tensor_and_offsets()` does not copy, which I believe it does not.

File: test/distributed/fsdp/test_fsdp_state_dict.py
Comment (rohan-varma) [reply test/distributed/fsdp/test_fsdp_state_dict.py:113]: https://github.com/pytorch/pytorch/issues/73044

File: torch/distributed/fsdp/flatten_params_wrapper.py
Comment (rohan-varma) [reply torch/distributed/fsdp/flatten_params_wrapper.py:434]: Is it worth to call `super().state_dict()` for consistency and avoid this confusion?

File: test/distributed/fsdp/test_fsdp_state_dict.py
Comment (rohan-varma) [reply test/distributed/fsdp/test_fsdp_state_dict.py:40]: https://github.com/pytorch/pytorch/issues/73048

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (fegin) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:793]: Fixed the inconsistency.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (fegin) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:761]: Fixed the inconsistency.

File: torch/distributed/fsdp/flatten_params_wrapper.py
Comment (fegin) [reply torch/distributed/fsdp/flatten_params_wrapper.py:434]: I removed this API since I changed the implementation.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (fegin) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:739]: No, the key will actually exist and we just replace the value.  I added comment to explain the logic.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (fegin) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:1279]: _rebuild_full_params are used in the several our tests. In the tests, there are no guarantee that forward will be called. Adding the lazy_init should make the code more robust.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (rohan-varma) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py]: great point!

**Review by rohan-varma — COMMENTED** [review]
Looks great! Couple of minor comments and wondering if we should also test `with state_dict_type` API in addition to direct call to `local_state_dict`.

Will stamp it after CI/lint fixes

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (fegin) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:861]: Not, we can abstract this in a follow up PR.

**Review by rohan-varma — APPROVED** [review]
LGTM assuming the context manager test is added, thanks for working on this! Please ensure CI is clear before landing

**Comment by suo** [comment]
Looks like this PR broke master: https://www.torch-ci.com/pytorch/pytorch/commit/d50643adcd8ca144c9b2d5de339b95be5339bf63

Unfortunately I have to revert

**Comment by fegin** [comment]
Interesting, this PR does not touch any files listed in the tests. So it should be impossible that this PR cause those flake8 and linter errors. @suo Just curious, did reverting the PR actually recover the master?

**Comment by suo** [comment]
the linter errors were caused by a different commit, which I also had to
revert. I believe this commit broke
https://github.com/pytorch/pytorch/runs/5260788165?check_suite_focus=true;
reverting did indeed fix master for that test

**Pull request closed without merging**
